File: swift/src/SHA256.swift
/// Pure Swift SHA-256 implementation
/// Parity: libgit2 EXPERIMENTAL_SHA256 uses SHA-256 for object IDs
import Foundation

public struct SHA256Hash {
    private var h0: UInt32 = 0x6a09e667
    private var h1: UInt32 = 0xbb67ae85
    private var h2: UInt32 = 0x3c6ef372
    private var h3: UInt32 = 0xa54ff53a
    private var h4: UInt32 = 0x510e527f
    private var h5: UInt32 = 0x9b05688c
    private var h6: UInt32 = 0x1f83d9ab
    private var h7: UInt32 = 0x5be0cd19

    private var buffer = [UInt8]()
    private var totalLength: UInt64 = 0

    private static let k: [UInt32] = [
        0x428a2f98, 0x71374491, 0xb5c0fbcf, 0xe9b5dba5,
        0x3956c25b, 0x59f111f1, 0x923f82a4, 0xab1c5ed5,
        0xd807aa98, 0x12835b01, 0x243185be, 0x550c7dc3,
        0x72be5d74, 0x80deb1fe, 0x9bdc06a7, 0xc19bf174,
        0xe49b69c1, 0xefbe4786, 0x0fc19dc6, 0x240ca1cc,
        0x2de92c6f, 0x4a7484aa, 0x5cb0a9dc, 0x76f988da,
        0x983e5152, 0xa831c66d, 0xb00327c8, 0xbf597fc7,
        0xc6e00bf3, 0xd5a79147, 0x06ca6351, 0x14292967,
        0x27b70a85, 0x2e1b2138, 0x4d2c6dfc, 0x53380d13,
        0x650a7354, 0x766a0abb, 0x81c2c92e, 0x92722c85,
        0xa2bfe8a1, 0xa81a664b, 0xc24b8b70, 0xc76c51a3,
        0xd192e819, 0xd6990624, 0xf40e3585, 0x106aa070,
        0x19a4c116, 0x1e376c08, 0x2748774c, 0x34b0bcb5,
        0x391c0cb3, 0x4ed8aa4a, 0x5b9cca4f, 0x682e6ff3,
        0x748f82ee, 0x78a5636f, 0x84c87814, 0x8cc70208,
        0x90befffa, 0xa4506ceb, 0xbef9a3f7, 0xc67178f2,
    ]

    public init() {}

    public mutating func update(_ data: [UInt8]) {
        buffer.append(contentsOf: data)
        totalLength += UInt64(data.count)

        while buffer.count >= 64 {
            let block = Array(buffer.prefix(64))
            processBlock(block)
            buffer.removeFirst(64)
        }
    }

    public mutating func update(_ string: String) {
        update(Array(string.utf8))
    }

    public mutating func finalize() -> [UInt8] {
        var padded = buffer
        padded.append(0x80)

        while padded.count % 64 != 56 {
            padded.append(0x00)
        }

        let bitLength = totalLength * 8
        for i in (0..<8).reversed() {
            padded.append(UInt8((bitLength >> (i * 8)) & 0xFF))
        }

        var offset = 0
        while offset < padded.count {
            let block = Array(padded[offset..<offset + 64])
            processBlock(block)
            offset += 64
        }

        var digest = [UInt8](repeating: 0, count: 32)
        for (i, h) in [h0, h1, h2, h3, h4, h5, h6, h7].enumerated() {
            digest[i * 4]     = UInt8((h >> 24) & 0xFF)
            digest[i * 4 + 1] = UInt8((h >> 16) & 0xFF)
            digest[i * 4 + 2] = UInt8((h >> 8) & 0xFF)
            digest[i * 4 + 3] = UInt8(h & 0xFF)
        }
        return digest
    }

    private mutating func processBlock(_ block: [UInt8]) {
        var w = [UInt32](repeating: 0, count: 64)

        for i in 0..<16 {
            w[i] = UInt32(block[i * 4]) << 24
                | UInt32(block[i * 4 + 1]) << 16
                | UInt32(block[i * 4 + 2]) << 8
                | UInt32(block[i * 4 + 3])
        }

        for i in 16..<64 {
            let s0 = rotateRight(w[i - 15], by: 7) ^ rotateRight(w[i - 15], by: 18) ^ (w[i - 15] >> 3)
            let s1 = rotateRight(w[i - 2], by: 17) ^ rotateRight(w[i - 2], by: 19) ^ (w[i - 2] >> 10)
            w[i] = w[i - 16] &+ s0 &+ w[i - 7] &+ s1
        }

        var a = h0, b = h1, c = h2, d = h3
        var e = h4, f = h5, g = h6, h = h7

        for i in 0..<64 {
            let s1 = rotateRight(e, by: 6) ^ rotateRight(e, by: 11) ^ rotateRight(e, by: 25)
            let ch = (e & f) ^ ((~e) & g)
            let temp1 = h &+ s1 &+ ch &+ Self.k[i] &+ w[i]
            let s0 = rotateRight(a, by: 2) ^ rotateRight(a, by: 13) ^ rotateRight(a, by: 22)
            let maj = (a & b) ^ (a & c) ^ (b & c)
            let temp2 = s0 &+ maj

            h = g
            g = f
            f = e
            e = d &+ temp1
            d = c
            c = b
            b = a
            a = temp1 &+ temp2
        }

        h0 = h0 &+ a
        h1 = h1 &+ b
        h2 = h2 &+ c
        h3 = h3 &+ d
        h4 = h4 &+ e
        h5 = h5 &+ f
        h6 = h6 &+ g
        h7 = h7 &+ h
    }

    private func rotateRight(_ value: UInt32, by count: UInt32) -> UInt32 {
        (value >> count) | (value << (32 - count))
    }

    /// Convenience: hash data in one call
    public static func hash(_ data: [UInt8]) -> [UInt8] {
        var sha = SHA256Hash()
        sha.update(data)
        return sha.finalize()
    }

    /// Convenience: hash string
    public static func hash(_ string: String) -> [UInt8] {
        hash(Array(string.utf8))
    }
}

// MARK: - Hash Algorithm

/// Hash algorithm selection (matching libgit2 EXPERIMENTAL_SHA256)
public enum HashAlgorithm: Sendable {
    case sha1
    case sha256

    /// Digest length in bytes
    public var digestLength: Int {
        switch self {
        case .sha1: return 20
        case .sha256: return 32
        }
    }

    /// Hex string length
    public var hexLength: Int {
        digestLength * 2
    }
}

// MARK: - OID SHA-256 Extensions

extension OID {
    /// SHA-256 digest length in bytes
    public static let sha256Length = 32

    /// SHA-256 hex string length
    public static let sha256HexLength = 64

    /// Create an OID by hashing data with SHA-256 (git object style, experimental)
    public static func hashSHA256(type: ObjectType, data: [UInt8]) -> OID {
        let typeName: String
        switch type {
        case .commit: typeName = "commit"
        case .tree:   typeName = "tree"
        case .blob:   typeName = "blob"
        case .tag:    typeName = "tag"
        }

        let header = "\(typeName) \(data.count)\0"
        var sha = SHA256Hash()
        sha.update(Array(header.utf8))
        sha.update(data)
        return OID(raw: sha.finalize())
    }

    /// Zero OID for SHA-256
    public static let zeroSHA256 = OID(raw: [UInt8](repeating: 0, count: sha256Length))
}

File: rust/src/sha256.rs
//! SHA-256 implementation using the `sha2` crate for hardware-accelerated hashing.
//! Parity: libgit2 EXPERIMENTAL_SHA256 uses SHA-256 for object IDs

use sha2_crate::{Sha256 as Sha256Impl, Digest};

pub struct SHA256 {
    hasher: Sha256Impl,
}

impl SHA256 {
    pub fn new() -> Self {
        Self { hasher: Sha256Impl::new() }
    }

    pub fn update(&mut self, data: &[u8]) {
        self.hasher.update(data);
    }

    pub fn finalize(self) -> [u8; 32] {
        self.hasher.finalize().into()
    }

    /// Hash data in one call
    pub fn hash(data: &[u8]) -> [u8; 32] {
        let mut sha = SHA256::new();
        sha.update(data);
        sha.finalize()
    }
}

impl Default for SHA256 {
    fn default() -> Self {
        Self::new()
    }
}

impl crate::OID {
    /// SHA-256 digest length in bytes
    pub const SHA256_LENGTH: usize = 32;

    /// SHA-256 hex string length
    pub const SHA256_HEX_LENGTH: usize = 64;

    /// Create an OID by hashing data with SHA-256 (git object style, experimental)
    pub fn hash_object_sha256(obj_type: crate::ObjectType, data: &[u8]) -> Self {
        let mut header_buf = [0u8; 32];
        let header_len = crate::sha1::build_object_header(obj_type, data.len(), &mut header_buf);
        let mut sha = SHA256::new();
        sha.update(&header_buf[..header_len]);
        sha.update(data);
        Self::from_bytes(sha.finalize().to_vec())
    }

    /// Zero OID for SHA-256
    pub fn zero_sha256() -> Self {
        Self::from_bytes(vec![0u8; Self::SHA256_LENGTH])
    }
}

/// Hash algorithm selection (matching libgit2 EXPERIMENTAL_SHA256)
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum HashAlgorithm {
    SHA1,
    SHA256,
}

impl HashAlgorithm {
    /// Digest length in bytes
    pub fn digest_length(&self) -> usize {
        match self {
            HashAlgorithm::SHA1 => 20,
            HashAlgorithm::SHA256 => 32,
        }
    }

    /// Hex string length
    pub fn hex_length(&self) -> usize {
        self.digest_length() * 2
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_sha256_empty() {
        let digest = SHA256::hash(b"");
        let hex: String = digest.iter().map(|b| format!("{:02x}", b)).collect();
        assert_eq!(hex, "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855");
    }

    #[test]
    fn test_sha256_hello() {
        let digest = SHA256::hash(b"hello");
        let hex: String = digest.iter().map(|b| format!("{:02x}", b)).collect();
        assert_eq!(hex, "2cf24dba5fb0a30e26e83b2ac5b9e29e1b161e5c1fa7425e73043362938b9824");
    }

    #[test]
    fn test_sha256_git_blob() {
        let data = b"hello\n";
        let oid = crate::OID::hash_object_sha256(crate::ObjectType::Blob, data);
        assert_eq!(oid.hex().len(), 64);
        assert!(!oid.is_zero());
    }

    #[test]
    fn test_sha256_longer() {
        let digest = SHA256::hash(b"The quick brown fox jumps over the lazy dog");
        let hex: String = digest.iter().map(|b| format!("{:02x}", b)).collect();
        assert_eq!(hex, "d7a8fbb307d7809469ca9abcb0082e4f8d5651e46d3cdb762d02d0bf37c9e592");
    }

    #[test]
    fn test_hash_algorithm() {
        assert_eq!(HashAlgorithm::SHA1.digest_length(), 20);
        assert_eq!(HashAlgorithm::SHA256.digest_length(), 32);
        assert_eq!(HashAlgorithm::SHA1.hex_length(), 40);
        assert_eq!(HashAlgorithm::SHA256.hex_length(), 64);
    }

    #[test]
    fn test_zero_sha256() {
        let z = crate::OID::zero_sha256();
        assert!(z.is_zero());
        assert_eq!(z.hex().len(), 64);
    }
}
